### PR TITLE
feat: add ruby code block - merge sort

### PR DESCRIPTION
**Pull request by junminhong**
If this pull request (PR) pertains to **Chinese-to-English translation**, please confirm that you have read the contribution guidelines and complete the checklist below:

- [ ] This PR represents the translation of a single, complete document, or contains only bug fixes.
- [ ] The translation accurately conveys the original meaning and intent of the Chinese version. If deviations exist, I have provided explanatory comments to clarify the reasons.

If this pull request (PR) is associated with **coding or code transpilation**, please attach the relevant console outputs to the PR and complete the following checklist:

- [x] I have thoroughly reviewed the code, focusing on its formatting, comments, indentation, and file headers.
- [x] I have confirmed that the code execution outputs are consistent with those produced by the reference code (Python or Java).
- [x] The code is designed to be compatible on standard operating systems, including Windows, macOS, and Ubuntu.

## result screenshot
Screenshot 2024-04-10 at 17 22 41

### CODE REVIEW

**Comment by krahets** [comment]
Hi @juminhong, please complete the check list.

File: codes/ruby/chapter_sorting/merge_sort.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  # 将临时数组tmp中的元素复制回原数组nums的对应区间
+  (0...tmp.length).each do |k|
+    nums[left + k] = tmp[k]
+  end
+end
+
+### 归并排序 ###
+def merge_sort(nums, left, right)
+  # 终止条件
+  # 当子数组长度为1时终止递归
```
Comment (krahets) [inline codes/ruby/chapter_sorting/merge_sort.rb:45]: Same as above

File: codes/ruby/chapter_sorting/merge_sort.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    tmp[k] = nums[i]
+    i += 1
+    k += 1
+  end
+  while j <= right
+    tmp[k] = nums[j]
+    j += 1
+    k += 1
+  end
+  # 将临时数组tmp中的元素复制回原数组nums的对应区间
```
Comment (krahets) [inline codes/ruby/chapter_sorting/merge_sort.rb:36]: Same as above

Comment (junminhong) [reply codes/ruby/chapter_sorting/merge_sort.rb]: [0a6dba9](https://github.com/krahets/hello-algo/pull/1260/commits/0a6dba9542a2b7253934253259fcfd9bf277bca6)

File: codes/ruby/chapter_sorting/merge_sort.rb (hunk 1 above)
Comment (junminhong) [reply codes/ruby/chapter_sorting/merge_sort.rb]: [0a6dba9](https://github.com/krahets/hello-algo/pull/1260/commits/0a6dba9542a2b7253934253259fcfd9bf277bca6)

File: codes/ruby/chapter_sorting/merge_sort.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
+=begin
+File: merge_sort.rb
+Created Time: 2024-04-10
+Author: junminhong (junminhong1110@gmail.com)
+=end
+
+### 合并左子数组和右子数组 ###
+def merge(nums, left, mid, right)
+  # 左子数组区间为 [left, mid], 右子数组区间为 [mid+1, right]
+  # 创建一个临时数组tmp，用于存放合并后的结果
```
[the remark this answers is not in the document]
Comment (junminhong) [reply codes/ruby/chapter_sorting/merge_sort.rb]: [0a6dba9](https://github.com/krahets/hello-algo/pull/1260/commits/0a6dba9542a2b7253934253259fcfd9bf277bca6)

File: codes/ruby/chapter_sorting/merge_sort.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+File: merge_sort.rb
+Created Time: 2024-04-10
+Author: junminhong (junminhong1110@gmail.com)
+=end
+
+### 合并左子数组和右子数组 ###
+def merge(nums, left, mid, right)
+  # 左子数组区间为 [left, mid], 右子数组区间为 [mid+1, right]
+  # 创建一个临时数组 tmp，用于存放合并后的结果
+  tmp = Array.new(right - left + 1)
```
Comment (khoaxuantu) [inline codes/ruby/chapter_sorting/merge_sort.rb:11]: It is better to preserve consistency with other codes. As in other codes, the `tmp` list is initialized as `[0, 0, 0, 0, 0]`

```suggestion
  tmp = Array.new(right - left + 1, 0)
```

**Review by khoaxuantu — COMMENTED** [review]
Hi @junminhong, I just have one small comment, the rest LGTM

File: codes/ruby/chapter_sorting/merge_sort.rb (hunk 4 above)
Comment (junminhong) [reply codes/ruby/chapter_sorting/merge_sort.rb]: thanks

Comment (junminhong) [reply codes/ruby/chapter_sorting/merge_sort.rb]: [7cf1610](https://github.com/krahets/hello-algo/pull/1260/commits/7cf1610445510cba0d3282c5e73d5dffdea08253)

File: codes/ruby/chapter_sorting/merge_sort.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+### 合并左子数组和右子数组 ###
+def merge(nums, left, mid, right)
+  # 左子数组区间为 [left, mid], 右子数组区间为 [mid+1, right]
+  # 创建一个临时数组 tmp，用于存放合并后的结果
+  tmp = Array.new(right - left + 1, 0)
+  # 初始化左子数组和右子数组的起始索引
+  i, j, k = left, mid + 1, 0
+  # 当左右子数组都还有元素时，进行比较并将较小的元素复制到临时数组中
+  while i <= mid and j <= right
```
Comment (khoaxuantu) [inline codes/ruby/chapter_sorting/merge_sort.rb:15]:
```suggestion
  while i <= mid && j <= right
```

File: codes/ruby/chapter_sorting/merge_sort.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  # 终止条件
+  # 当子数组长度为 1 时终止递归
+  return if left >= right
+  # 划分阶段
+  # 计算中点
+  mid = (left + right) / 2
+  # 递归左子数组
+  merge_sort(nums, left, mid)
+  # 递归右子数组
+  merge_sort(nums, mid + 1, right)
```
Comment (khoaxuantu) [inline codes/ruby/chapter_sorting/merge_sort.rb:53]: I think we should maintain the consistency with other codes.

```suggestion
  mid = (left + right) / 2 # 计算中点
  merge_sort(nums, left, mid) # 递归左子数组
  merge_sort(nums, mid + 1, right) # 递归右子数组
```

File: codes/ruby/chapter_sorting/merge_sort.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  # 合并阶段
+  merge(nums, left, mid, right)
+end
+
+### Driver Code ###
+nums = [7, 3, 2, 6, 0, 1, 5, 4]
+merge_sort(nums, 0, nums.length - 1)
+puts "归并排序完成后 nums = #{nums.inspect}"
```
Comment (khoaxuantu) [inline codes/ruby/chapter_sorting/merge_sort.rb:61]: Please help me add the driver code into the `if __FILE__ == $0` block, we just made this change recently.

```suggestion
### Driver Code ###
if __FILE__ == $0
  nums = [7, 3, 2, 6, 0, 1, 5, 4]
  merge_sort(nums, 0, nums.length - 1)
  puts "归并排序完成后 nums = #{nums.inspect}"
end
```

**Review by khoaxuantu — COMMENTED** [review]
Hi @junminhong, could you help me change the code as the following:

**Review by khoaxuantu — APPROVED** [review]
LGTM

**Review by krahets — APPROVED** [review]
Thanks!

**Pull request merged**